### PR TITLE
VTOL rate control improvements

### PR DESCRIPTION
**Pull request by RomanBapst**
Replaces #9190

### CODE REVIEW

**Comment by LorenzMeier** [comment]
Not merging just yet!

**Comment by bkueng** [comment]
@LorenzMeier yes, please don't! It breaks Flight Review :)

File: src/modules/fw_att_control/FixedwingAttitudeControl.cpp [hunk 1]
```diff
@@ -335,14 +341,38 @@ FixedwingAttitudeControl::vehicle_manual_poll()
...
-FixedwingAttitudeControl::vehicle_setpoint_poll()
+FixedwingAttitudeControl::vehicle_attitude_setpoint_poll()
 {
 	/* check if there is a new setpoint */
 	bool att_sp_updated;
 	orb_check(_att_sp_sub, &att_sp_updated);
 
 	if (att_sp_updated) {
-		orb_copy(ORB_ID(vehicle_attitude_setpoint), _att_sp_sub, &_att_sp);
+		if (orb_copy(ORB_ID(vehicle_attitude_setpoint), _att_sp_sub, &_att_sp)) {
```
Comment (bkueng) [inline src/modules/fw_att_control/FixedwingAttitudeControl.cpp]:
```suggestion
		if (orb_copy(ORB_ID(vehicle_attitude_setpoint), _att_sp_sub, &_att_sp) == 0) {
```

File: src/modules/fw_att_control/FixedwingAttitudeControl.cpp [hunk 2]
```diff
@@ -404,6 +443,29 @@ FixedwingAttitudeControl::vehicle_land_detected_poll()
...
+	}
+
+	/*
+	 * For scaling our actuators using anything less than the min (close to stall)
+	 * speed doesn't make any sense - its the strongest reasonable deflection we
+	 * want to do in flight and its the baseline a human pilot would choose.
+	 *
+	 * Forcing the scaling to this value allows reasonable handheld tests.
+	 */
+	return _parameters.airspeed_trim / ((airspeed < _parameters.airspeed_min) ? _parameters.airspeed_min :
```
Comment (bkueng) [inline src/modules/fw_att_control/FixedwingAttitudeControl.cpp]:
```suggestion
	return _parameters.airspeed_trim / math::max(airspeed,  _parameters.airspeed_min);
```

File: src/modules/mc_att_control/mc_att_control_main.cpp [hunk 3]
```diff
@@ -784,7 +785,6 @@ MulticopterAttitudeControl::run()
 				if (!_vehicle_status.is_vtol) {
 					_rates_sp.zero();
 					_rates_int.zero();
-					_thrust_sp = 0.0f;
```
Comment (bkueng) [inline src/modules/mc_att_control/mc_att_control_main.cpp]: Why is this removed?

File: src/modules/vtol_att_control/tailsitter.cpp [hunk 4]
```diff
@@ -176,89 +177,91 @@ void Tailsitter::update_transition_state()
...
+			// attitude during transitions are controlled by mc attitude control so rotate the desired attitude to the
+			// multirotor frame
+			_q_trans_start = _q_trans_start * matrix::Quatf(matrix::Eulerf(0, -M_PI_2_F, 0));
 
-		} else {
-			_mc_yaw_weight = 1.0f;
+		} else if (_vtol_schedule.flight_mode == TRANSITION_FRONT_P1) {
+			// initial attitude setpoint for the transition should be with wings level
+			_q_trans_start = matrix::Eulerf(0.0f, _mc_virtual_att_sp->pitch_body, _mc_virtual_att_sp->yaw_body);
+			matrix::Vector3f x = matrix::Dcmf(matrix::Quatf(_v_att->q)) * matrix::Vector3f(1, 0, 0);
```
Comment (bkueng) [inline src/modules/vtol_att_control/tailsitter.cpp]: Can you add a `using namespace matrix` on top? Should become more readable w/o `matrix::` everywhere.

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 5]
```diff
@@ -1448,7 +1448,7 @@ MavlinkReceiver::handle_message_set_attitude_target(mavlink_message_t *msg)
 					}
 
 					if (!_offboard_control_mode.ignore_thrust) { // dont't overwrite thrust if it's invalid
-						att_sp.thrust = set_attitude_target.thrust;
+						att_sp.thrust_x = set_attitude_target.thrust;
```
Comment (bkueng) [inline src/modules/mavlink/mavlink_receiver.cpp]: Is x here correct? Further down it's using z.

File: src/modules/mc_att_control/mc_att_control_main.cpp [hunk 6]
```diff
@@ -460,7 +459,7 @@ Vector3f
 MulticopterAttitudeControl::pid_attenuations(float tpa_breakpoint, float tpa_rate)
 {
 	/* throttle pid attenuation factor */
-	float tpa = 1.0f - tpa_rate * (fabsf(_v_rates_sp.thrust) - tpa_breakpoint) / (1.0f - tpa_breakpoint);
+	float tpa = 1.0f - tpa_rate * (fabsf(-_v_rates_sp.thrust_z) - tpa_breakpoint) / (1.0f - tpa_breakpoint);
```
Comment (RomanBapst) [inline src/modules/mc_att_control/mc_att_control_main.cpp]: Note to myself: change to _thrust_sp

File: src/modules/mc_att_control/mc_att_control_main.cpp (hunk 3 above)
Comment (RomanBapst) [reply src/modules/mc_att_control/mc_att_control_main.cpp]: @bkueng Mistake during rebase

File: src/modules/vtol_att_control/tailsitter.cpp (hunk 4 above)
Comment (RomanBapst) [reply src/modules/vtol_att_control/tailsitter.cpp]: @bkueng Sure, good idea as it's not a header file.

File: src/modules/mc_att_control/mc_att_control_main.cpp (hunk 3 above)
Comment (RomanBapst) [reply src/modules/mc_att_control/mc_att_control_main.cpp]: Fixed

File: src/modules/vtol_att_control/tailsitter.cpp (hunk 4 above)
Comment (RomanBapst) [reply src/modules/vtol_att_control/tailsitter.cpp]: Done

File: src/modules/mavlink/mavlink_receiver.cpp (hunk 5 above)
Comment (RomanBapst) [reply src/modules/mavlink/mavlink_receiver.cpp]: Fixed, for now use thrust_z as offboard is used for multicopters

File: src/modules/mc_att_control/mc_att_control_main.cpp (hunk 6 above)
Comment (RomanBapst) [reply src/modules/mc_att_control/mc_att_control_main.cpp]: Fixed

**Comment by RomanBapst** [comment]
@PX4/testflights Any chance you can test fly this. It touches multirotor, vtol and fixed wing.

File: src/modules/navigator/gpsfailure.cpp [hunk 7]
```diff
@@ -86,7 +86,7 @@ GpsFailure::on_active()
 			att_sp.timestamp = hrt_absolute_time();
 			att_sp.roll_body = math::radians(_param_openlooploiter_roll.get());
 			att_sp.pitch_body = math::radians(_param_openlooploiter_pitch.get());
-			att_sp.thrust = _param_openlooploiter_thrust.get();
+			att_sp.thrust_x = _param_openlooploiter_thrust.get();
```
Comment (MaEtUgR) [inline src/modules/navigator/gpsfailure.cpp]: It seems obvious that this is only done in fixed wing mode but where is that decision?

File: msg/vehicle_attitude_setpoint.msg [hunk 8]
```diff
@@ -12,7 +12,11 @@ float32 yaw_sp_move_rate	# rad/s (commanded by user)
 float32[4] q_d			# Desired quaternion for quaternion control
 bool q_d_valid			# Set to true if quaternion vector is valid
 
-float32 thrust			# Thrust in Newton the power system should generate
+# For clarification: For multicopters thrust_x and thrust_y are usually 0 and thrust_z is the negative throttle demand.
+# For fixed wings thrust_x is the throttle demand and thrust_y, thrust_z will usually be zero.
+float32 thrust_x					# Throttle command in body x direction [-1,1]
```
Comment (MaEtUgR) [inline msg/vehicle_attitude_setpoint.msg]: I strongly suggest that we use arrays to store vecotrs inside uORB messages like here:
https://github.com/PX4/Firmware/blob/master/msg/vehicle_local_position_setpoint.msg#L20

File: src/modules/navigator/gpsfailure.cpp (hunk 7 above)
Comment (RomanBapst) [reply src/modules/navigator/gpsfailure.cpp]: @MaEtUgR The gpsfailure class is only used for fixed wing. It's executed if the navigation state is set to NAVIGATION_STATE_AUTO_LANDGPSFAIL. If you go look into the commander you'll see that it's only used for fixed wing.

File: msg/vehicle_attitude_setpoint.msg (hunk 8 above)
Comment (RomanBapst) [reply msg/vehicle_attitude_setpoint.msg]: @MaEtUgR Done!

File: src/modules/fw_att_control/FixedwingAttitudeControl.cpp [hunk 9]
```diff
@@ -249,22 +285,23 @@ FixedwingAttitudeControl::parameters_update()
...
 	}
 }
 
 void
 FixedwingAttitudeControl::vehicle_manual_poll()
 {
-	// only update manual if in a manual mode
-	if (_vcontrol_mode.flag_control_manual_enabled) {
+	bool manual_updated = false;
+	orb_check(_manual_sub, &manual_updated);
```
Comment (dagar) [inline src/modules/fw_att_control/FixedwingAttitudeControl.cpp]: @RomanBapst I believe this change was done intentionally for flaps deployment. I'll track down the original so we can make sure the intended fix is preserved.

Comment (dagar) [reply src/modules/fw_att_control/FixedwingAttitudeControl.cpp]: https://github.com/PX4/Firmware/pull/9607

**Comment by dagar** [comment]
@PX4/testflights please try this on all VTOLs, as well as regression testing on MC and FW platforms.

**Comment by dannyfpv** [comment]
test on pixhawk 1 and crash on stabilize mode
https://review.px4.io/plot_app?log=8f5cbf75-cb1f-4fd3-88ed-d18d425c7cfb

on pixhawk v4 pro crash on stabilize mode
https://review.px4.io/plot_app?log=014173eb-4156-4828-a97d-492c417f3577

**Comment by RomanBapst** [comment]
@dannyfpv Hi! Thanks for testing and sorry for the crash. I checked the log and I already found the bug. Unfortunately, there was sign error in throttle in stabilized mode that was introduced during the lasted rebase on master. SITL did not catch that bug because it does not use test stabilized mode.
However, in future I suggest to bench test PRs which introduce massive changes like this one does before actually flying the vehicle. In this case you would have had no throttle in stabilized mode.
Anyway, I hope the convergence is not broken too badly...

**Comment by mrpollo** [comment]
Thanks for the feedback @RomanBapst, don't worry about the convergence it fulfilled its duty! Now let's make sure we have tests for stabilized mode, should I open an issue for tracking @dagar or is this covered somewhere already?

**Comment by RomanBapst** [comment]
@mrpollo I was already thinking about this. I think it would not be too hard to use the dronecode SDK in the CI and actually 'simulate' a human flying with a radio controller. That would just be a sluggish controller outputting RC commands.

File: msg/vehicle_attitude_setpoint.msg (hunk 8 above)
Comment (MaEtUgR) [reply msg/vehicle_attitude_setpoint.msg]: Perfect thanks!

**Comment by dannyfpv** [comment]
tested on pixhawk v4 pro
https://review.px4.io/plot_app?log=482daec6-d8b5-416f-bb3c-e57782fec3bd
https://review.px4.io/plot_app?log=4fc131d3-e215-4188-895c-d59a7161747d

**Comment by dagar** [comment]
I took another pass and this looks good. Let's merge and keep going.

**Comment by mrpollo** [comment]
Great team work everyone 🎉
